test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java

```diff
@@ -0,0 +1,408 @@
+                return;
+            default:
+                throw new UnsupportedOperationException("Unknown action type: " + action);
+        }
+        if (cb != null)
+        {
+            scheduler.schedule(() -> {
+                CallbackContext ctx = callbacks.remove(new CallbackKey(message.id(), to));
+                if (ctx != null)
+                {
+                    assert ctx == cb;
+                    try
+                    {
+                        ctx.onFailure(to, RequestFailureReason.TIMEOUT);
+                    }
+                    catch (Throwable t)
+                    {
+                        onError.accept(t);
+                    }
+                }
+            }, message.verb().expiresAfterNanos(), TimeUnit.NANOSECONDS);
+        }
+    }
+
+    private void deliver(Message<?> message, InetAddressAndPort to)
+    {
+        Duration delay = networkDelay.jitter(message, to);
+        if (delay == null) reciever.accept(to, message);
+        else               scheduler.schedule(() -> reciever.accept(to, message), delay.toNanos(), TimeUnit.NANOSECONDS);
+    }
+
+    @SuppressWarnings("rawtypes")
+    public SimulatedMessageReceiver reciver(IVerbHandler onMessage)
+    {
+        return new SimulatedMessageReceiver(onMessage);
+    }
+
+    public class SimulatedMessageReceiver
+    {
+        @SuppressWarnings("rawtypes")
+        final IVerbHandler onMessage;
+
+        @SuppressWarnings("rawtypes")
+        public SimulatedMessageReceiver(IVerbHandler onMessage)
+        {
+            this.onMessage = onMessage;
+        }
+
+        public void recieve(Message<?> msg)
+        {
+            if (status != Status.Up)
+                return;
+            if (msg.verb().isResponse())
+            {
+                CallbackKey key = new CallbackKey(msg.id(), msg.from());
+                if (callbacks.containsKey(key))
+                {
+                    CallbackContext callback = callbacks.remove(key);
+                    if (callback == null)
+                        return;
+                    try
+                    {
+                        if (msg.isFailureResponse())
+                            callback.onFailure(msg.from(), (RequestFailureReason) msg.payload);
+                        else callback.onResponse(msg);
+                    }
+                    catch (Throwable t)
+                    {
+                        onError.accept(t);
+                    }
+                }
+            }
+            else
+            {
+                try
+                {
+                    //noinspection unchecked
+                    onMessage.doVerb(msg);
+                }
+                catch (Throwable t)
+                {
+                    onError.accept(t);
+                }
+            }
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    public static class SimpleVerbHandler implements IVerbHandler
```

Comment:
nit: unused now. Should we keep this for future usages?

Reply:
its used in https://issues.apache.org/jira/browse/CASSANDRA-19769 (where this was forked from).  Right now a single test uses this outer class and doesn't deal with multiple verbs, but CASSANDRA-19769 has more examples that do... I "could" refactor repair to use this class (as its a lot of copy/paste) but wanted to keep the diff small...

I do know our style guide says to not add dead code... but feel that this is useful so others can start using and not have to do as much work.